odl/solvers/nonsmooth/admm.py

```diff
@@ -0,0 +1,155 @@
+
+    Notes
+    -----
+    The ADMM method applies the following iteration:
+
+    .. math::
+        x^{(k+1)} &= \mathrm{prox}_{\\tau f} \\left[
+            x^{(k)} - \sigma^{-1}\\tau L^*\\big(
+                L x^{(k)} - z^{(k)} + u^{(k)}
+            \\big)
+        \\right]
+
+        z^{(k+1)} &= \mathrm{prox}_{\sigma g}\\left(
+            L x^{(k+1)} + u^{(k)}
+        \\right)
+
+        u^{(k+1)} &= u^{(k)} + L x^{(k+1)} - z^{(k+1)}
+
+    The step size parameters :math:`\\tau` and :math:`\sigma` must satisfy
+
+    .. math::
+        0 < \\tau < \\frac{\sigma}{\|L\|^2}
+
+    to guarantee convergence.
+
+    The name "linearized ADMM" comes from the fact that in the
+    minimization subproblem for the :math:`x` variable, this variant
+    uses a linearization of a quadratic term in the augmented Lagrangian
+    of the generic ADMM, in order to make the step expressible with
+    the proximal operator of :math:`f`.
+
+    Another name for this algorithm is *split inexact Uzawa method*.
+
+    References
+    ----------
+    [PB2014] Parikh, N and Boyd, S. *Proximal Algorithms*. Foundations and
+    Trends in Optimization, 1(3) (2014), pp 123-231.
+    """
+    if not isinstance(L, Operator):
+        raise TypeError('`op` {!r} is not an `Operator` instance'
+                        ''.format(L))
+
+    if x not in L.domain:
+        raise OpDomainError('`x` {!r} is not in the domain of `op` {!r}'
+                            ''.format(x, L.domain))
+
+    tau, tau_in = float(tau), tau
+    if tau <= 0:
+        raise ValueError('`tau` must be positive, got {}'.format(tau_in))
+
+    sigma, sigma_in = float(sigma), sigma
+    if sigma <= 0:
+        raise ValueError('`sigma` must be positive, got {}'.format(sigma_in))
+
+    niter, niter_in = int(niter), niter
+    if niter < 0 or niter != niter_in:
+        raise ValueError('`niter` must be a non-negative integer, got {}'
+                         ''.format(niter_in))
+
+    # Callback object
+    callback = kwargs.pop('callback', None)
+    if callback is not None and not callable(callback):
+        raise TypeError('`callback` {} is not callable'.format(callback))
+
+    # Initialize range variables
+    z = L.range.zero()
+    u = L.range.zero()
+
+    # Temporary for Lx + u [- z]
+    tmp_ran = L(x)
+    # Temporary for L^*(Lx + u - z)
+    tmp_dom = L.domain.element()
+
+    for _ in range(niter):
+        # tmp_ran has value L x^k here
+        # tmp_dom <- L^*(L x^k + u^k - z^k)
+        tmp_ran += u
+        tmp_ran -= z
+        L.adjoint(tmp_ran, out=tmp_dom)
+
+        # x <- x^k - (tau/sigma) L^*(Lx^k + u^k - z^k)
+        x.lincomb(1, x, -tau / sigma, tmp_dom)
+        # x^(k+1) <- prox[tau*f](x)
+        f.proximal(tau)(x, out=x)
+
+        # tmp_ran <- L x^(k+1)
+        L(x, out=tmp_ran)
+        # z^(k+1) <- prox[sigma*g](L x^(k+1) + u^k)
+        g.proximal(sigma)(tmp_ran + u, out=z)  # 1 copy here
```

Comment:
why not `tmp_ran += u` on the line above?

Reply:
I need `L(x)` later to add the new `u`. The alternative would be to use another temporary, but that seems overkill to me.

Reply:
Go with this for now